src/fluid/con2prim_robust.hpp

```diff
@@ -398,26 +394,42 @@ class ConToPrim {
       SPACELOOP(i) { bu[i] = v(pb_lo + i); }
     }
 
-    Real ye_prim = 0.0;
-    if (pye > 0) ye_prim = v(pye);
-    Real ye_cons;
-    Real S[3];
-    Real bcons[3];
     Real sig[3];
-    prim2con::p2c(v(prho), vel, bu, v(peng), ye_prim, v(prs), v(gm1), g.gcov4, g.gcon,
-                  g.beta, g.lapse, g.gdet, v(crho), S, bcons, v(ceng), ye_cons, sig);
-
-    SPACELOOP(i) { v(cmom_lo + i) = S[i]; }
+    if (res.used_bounds()) {
+      Real ye_prim = 0.0;
+      if (pye > 0) ye_prim = v(pye);
+      Real ye_cons;
+      Real S[3];
+      Real bcons[3];
+      prim2con::p2c(v(prho), vel, bu, v(peng), ye_prim, v(prs), v(gm1), g.gcov4, g.gcon,
+                    g.beta, g.lapse, g.gdet, v(crho), S, bcons, v(ceng), ye_cons, sig);
+      SPACELOOP(i) { v(cmom_lo + i) = S[i]; }
+
+      if (pye > 0) v(cye) = ye_cons;
+    } else {
+      // just compute signal speeds
+      SPACELOOP(m) vel[m] *= iW;
+      bsq = 0.0;
+      if (pb_hi > 0) {
+        Real bdotv = 0.0;
+        SPACELOOP2(m, n) {
+          bsq += bu[m] * bu[n] * g.gcov4[m + 1][n + 1];
+          bdotv += bu[m] * vel[n] * g.gcov4[m + 1][n + 1];
+        }
+        bsq = iW * iW * bsq + bdotv * bdotv;
+      }
 
-    if (pye > 0) v(cye) = ye_cons;
+      prim2con::signal_speeds(v(prho), v(peng), v(prs), bsq, vel, vsq, v(gm1), g.lapse,
+                              g.beta, g.gcon, sig);
+    }
 
     for (int i = 0; i < sig_hi - sig_lo + 1; i++) {
       v(sig_lo + i) = sig[i];
     }
 
     num_nans = std::isnan(v(crho)) + std::isnan(v(cmom_lo)) + std::isnan(v(ceng));
 
-    if (num_nans > 0) {
+    if (num_nans > 0 || res.used_gamma_max()) {
```

Comment:
We may want to revisit this later to decide if `used_gamma_max()` really warrants a stencil fixup... but it's kind of an extreme failure that algorithms like this hate and the non-stencil solution is to basically shave off the maximum anyway so this may be fine. Just mentioning it.

Reply:
yeah, I think it's still unclear exactly when we want to say the inversion failed.  I imagine we'll figure this out as we start to run more production models.